unicon_runner/runner/task/programming.py

```diff
@@ -107,9 +141,40 @@ class ProgrammingTaskExpectedAnswer(BaseModel):
     expected_answer: Any
 
 
+class Link(BaseModel):
+    id: int
+    from_node_id: int
+    from_socket_id: int
+    to_node_id: int
+    to_socket_id: int
+
+
+# class LoopStep(Step):
+#     subgraph: Graph
+
+#     def set_user_input(self, user_input: list[File]):
+#         self.user_input = user_input
+
+#     async def run(
+#         self,
+#         inputs: dict,
+#         environment: ProgrammingEnvironment,
+#         executor: Executor,
+#     ):
+#         state = inputs.get("state")
+#         while True:
+#             outputs = await self.subgraph.run(state, environment, executor)
+#             if outputs["end"]:
+#                 break
+#         return outputs
+
+
 class Testcase(BaseModel):
     id: int
+
+    # Steps are assumed to be in topological order
```

Comment:
topological order is guaranteed by https://github.com/uniconhq/unicon-backend/pull/20